### PR TITLE
Fix stale leaf vlog after compaction

### PR DESCRIPTION
**Pull request by snissn**
## Objective

Fix stale split `leaf_vlog` retention after offline index vacuum and route final-footprint benchmark maintenance through the high-level TreeDB storage compaction path.

## Context

A benchmark run showed `maindb/leaf_vlog/value-l255-000001.log` remaining at about 19 MiB after full maintenance, while gzip compressed it to about 4 MiB. Investigation found that offline index vacuum rebuilds roots into pager-backed `index.db` leaves, but left the old writable split `leaf_vlog` generation on disk. The benchmark `full` maintenance path also hand-chained low-level rewrite/GC/leaf-pack/vacuum primitives instead of using `CompactStorage`.

## Non-goals

- This PR does not change live foreground leaf-log write encoding.
- This PR does not remove low-level maintenance APIs or diagnostic benchmark rows that intentionally measure individual primitives.
- This PR does not add new compression codecs.

## Design

- Reset split leaf-generation files, `.lenidx` sidecars, and manifest after successful `VacuumIndexOffline` swaps in a pager-backed rebuilt `index.db`.
- Change `mongo_gateway_bench -treedb-maintenance full` to run `CompactStorage` first, then close and run offline index vacuum for final-footprint shrink.
- Route application-db comparison scripts and the unified benchmark after-run storage report through `CompactStorage` instead of treating value-log rewrite as final compaction.
- Update scripts/docs that were using value-log rewrite as generic “compaction” guidance to point final-footprint workflows at `CompactStorage` / `treemap compact`.

## Scope

- Includes (files/symbols): `TreeDB/db/vacuum_offline.go`, `TestVacuumIndexOffline_LeafPagesInValueLog_ReopenParity`, `cmd/mongo_gateway_bench` full maintenance, `cmd/unified_bench` after-run storage reporting, application-db comparison scripts, canonical collection benchmark phase labels, `scripts/treedb_collection_compression_matrix.sh`, related docs.
- Excludes (files/symbols): live `leaf_vlog` append batching, low-level `ValueLogRewrite*`, `LeafGenerationPack*`, and scheduler internals.

## Correctness

- Tests run (exact commands):
  - `go test ./TreeDB -run TestVacuumIndexOffline_LeafPagesInValueLog_ReopenParity`
  - `go test ./TreeDB/db -run 'TestVacuumIndexOffline|TestLeafGenerationManifest|TestLeafGenerationGC'`
  - `go test ./cmd/collection_load_fixture -run 'TestRunFixtureLeafGenerationReportsDiskTotals|TestVacuumIndexOfflinePreservesTemplateV1TwoIndexDatabase' -count=1`
  - `go test ./TreeDB/...`
  - `go test ./cmd/mongo_gateway_bench`
  - `go test ./cmd/unified_bench`
  - `go test ./cmd/collection_canonical_bench`
  - `bash -n scripts/compare_application_db_rebuilds.sh scripts/run_application_db_engine_matrix.sh scripts/treedb_collection_compression_matrix.sh`
  - `go run ./cmd/mongo_gateway_bench -target treedb -documents 200 -batch-size 100 -reads 0 -range-reads 0 -updates 0 -secondary-indexes 2 -treedb-maintenance full -timeout 0 -format json`
- Corruption/edge cases covered: offline vacuum reopen parity now asserts stale `leaf_vlog` segment files are removed and reads still succeed after reopen.
- Concurrency/race coverage (if applicable): cleanup runs only after offline vacuum has closed the DB and completed the index swap.

## Performance

- Benchmarks run (exact commands): not run; this is a correctness/maintenance-path fix.
- Results table (before/after; median-of-5 per G1 in `docs/OPT_SPRINT_NEXT.md`): N/A.
- Regressions (if any) and why acceptable: `-treedb-maintenance full` now runs the fuller high-level path, so it may do more maintenance work than the previous bespoke chain. That is intentional for final storage-footprint reporting.

## Rollout / Toggle Plan

- Default setting: `mongo_gateway_bench` already defaults to `-treedb-maintenance full`; it now uses the recommended path.
- How to disable quickly: use `-treedb-maintenance checkpoint` or `-treedb-maintenance none` in the benchmark.

## Follow-ups

- Historical benchmark reports still mention the old value-log rewrite workflows as historical measurements; this PR does not rewrite past result artifacts.

---

### Scope
- [x] Single, clear objective for this PR (not a bundle)
- [x] Explicit include list (files/symbols touched): listed above
- [x] Explicit exclude list (files/symbols NOT touched): listed above
- [x] No unwanted feature reintroduction (e.g. async slab writer, hard zones, ValueIndex) unless explicitly stated in the PR objective

### Safety / Correctness
- [ ] Clear written-vs-durable semantics for any `*Sync` path touched
- [x] No new mmap usage on mutable/truncating files
- [ ] All new length fields are capped before allocation (OOM-safe)
- [ ] CRC/checksum behavior is fail-closed (clean error, no panic)
- [x] Any concurrency change has a defined state machine and invariants

### Tests
- [x] Added deterministic regression tests for the targeted failure mode
- [x] Tests avoid sleeps where possible (use latches/hooks)
- [ ] Added corruption/edge-case tests (truncation, bad headers, invalid lengths) when format/parsing changes
- [x] Ran locally:
  - [ ] `go test ./... -count=1`
  - [x] Any targeted packages/benchmarks: listed above

### Benchmarks / Measurements
- [ ] Added or updated a benchmark relevant to the change
- [ ] Reported results in PR description (before/after, command, machine)
- [ ] Included “incompressible” (worst-case) results if compression is involved

### Docs
- [ ] Updated `docs/OPT_SPRINT_NEXT.md` milestone status (if applicable)
- [ ] Documented any new config knobs + defaults
- [x] Called out any format change in PR description (pre-alpha OK): no format change

### Rollout / Toggle Plan (if behavior-affecting)
- [x] Safe defaults (feature off or conservative threshold)
- [x] Clear knobs to disable/revert behavior quickly
- [ ] Observability: counters/logs to verify effectiveness

### CODE REVIEW

File: TreeDB/db/vacuum_offline.go [hunk 1]
```diff
@@ -212,6 +212,72 @@ func vacuumIndexOffline(opts Options, fail vacuumFailpoint) error {
...
+	}
+	if len(files) == 0 && !manifestExists && len(sidecarPaths) == 0 {
+		return nil
+	}
+
+	for _, file := range files {
+		path := leafGenerationFallbackPath(dir, file.rawFileID)
+		if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
+			return fmt.Errorf("vacuum: remove stale leaf generation file %s: %w", path, err)
+		}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/vacuum_offline.go:364]: Fixed in dec23419 and still present in c634e230: the stale manifest is removed and the directory is synced before segment/sidecar deletion, then a reset manifest is written and synced after cleanup.

File: cmd/mongo_gateway_bench/main.go [hunk 2]
```diff
@@ -4646,6 +4612,67 @@ func runTreeDBMaintenanceStack(ctx context.Context, target *benchTarget, result
...
+}
+
+func compactStorageSkipReason(stats backenddb.CompactStorageStats) string {
+	if stats.FullyCompacted {
+		return ""
+	}
+	if stats.RemainingDebt.Empty() {
+		return ""
+	}
+	return "remaining_debt"
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/mongo_gateway_bench/main.go]: Fixed in dec23419: completed compaction with remaining debt no longer reports Skipped=remaining_debt; the metrics carry remaining debt instead.

File: TreeDB/db/vacuum_offline.go [hunk 3]
```diff
@@ -212,6 +212,77 @@ func vacuumIndexOffline(opts Options, fail vacuumFailpoint) error {
 		}
 	}
 
+	if !opts.IndexOuterLeavesInValueLog {
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/vacuum_offline.go]: Fixed in c634e230: offline vacuum now scans the rebuilt output roots for live leaf-log refs and resets leaf_vlog only when the rebuilt index no longer references leaf-log children.

File: TreeDB/db/vacuum_offline.go [hunk 4]
```diff
@@ -212,6 +212,77 @@ func vacuumIndexOffline(opts Options, fail vacuumFailpoint) error {
...
+		if err := resetLeafGenerationAfterOfflineVacuum(opts.Dir, meta.CommitSeq); err != nil {
+			return err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/vacuum_offline.go]: Fixed in c634e230: the cleanup is no longer gated on IndexOuterLeavesInValueLog; it is gated on structural live-ref detection in the rebuilt output.

File: TreeDB/maintenance_leaf_vlog_test.go [hunk 5]
```diff
@@ -185,10 +185,27 @@ func TestVacuumIndexOffline_LeafPagesInValueLog_ReopenParity(t *testing.T) {
...
 	if err := treedb.VacuumIndexOffline(treedb.Options{Dir: dir, KeepRecent: 1}); err != nil {
 		t.Fatalf("VacuumIndexOffline: %v", err)
 	}
 
+	leafPathsAfter, err := filepath.Glob(filepath.Join(mainDir, "leaf_vlog", "value-l*.log"))
+	if err != nil {
+		t.Fatalf("glob leaf_vlog after vacuum: %v", err)
+	}
+	if len(leafPathsAfter) == 0 {
+		t.Fatalf("offline vacuum removed live leaf_vlog files for outer-leaf storage")
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/maintenance_leaf_vlog_test.go]: Addressed in c634e230: the reopen-parity test no longer assumes leaf_vlog files must remain; it verifies correctness after vacuum while structural cleanup decides whether files are live or stale.

**Comment by snissn** [comment]
@codex review

File: TreeDB/db/vacuum_offline.go [hunk 6]
```diff
@@ -212,6 +218,145 @@ func vacuumIndexOffline(opts Options, fail vacuumFailpoint) error {
...
+	if len(files) == 0 && !manifestExists && len(sidecarPaths) == 0 {
+		return nil
+	}
+	if manifestExists {
+		manifestPath := leafGenerationManifestPath(leafDir)
+		if err := os.Remove(manifestPath); err != nil && !os.IsNotExist(err) {
+			return fmt.Errorf("vacuum: remove stale leaf generation manifest %s: %w", manifestPath, err)
+		}
+		if err := syncDirFn(leafDir); err != nil {
+			return fmt.Errorf("vacuum: sync leaf_vlog dir after manifest removal: %w", err)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/vacuum_offline.go]: Fixed in f72ef5ea: reset now writes and syncs the empty leaf-generation manifest before deleting stale leaf-log segment files, and the new regression test injects a stop after the reset manifest to prove recovery will not bootstrap the stale files again.

**Comment by snissn** [comment]
@codex review

File: TreeDB/db/vacuum_offline.go [hunk 7]
```diff
@@ -212,6 +218,135 @@ func vacuumIndexOffline(opts Options, fail vacuumFailpoint) error {
 		}
 	}
 
+	if !outputHasLeafLogRefs {
+		if err := resetLeafGenerationAfterOfflineVacuum(opts.Dir, meta.CommitSeq); err != nil {
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/vacuum_offline.go]: Fixed in 4552c16b: read-write open now scans the current roots for leaf-log refs and resets stale leaf-generation state before loading the manifest when no live leaf-log refs remain. Added a regression test that recreates the post-swap stale manifest/files state and verifies reopen clears them.

File: TreeDB/maintenance_leaf_vlog_test.go [hunk 8]
```diff
@@ -185,6 +185,15 @@ func TestVacuumIndexOffline_LeafPagesInValueLog_ReopenParity(t *testing.T) {
...
 	}
 
+	mainDir := filepath.Dir(mainIndexPath(dir))
+	leafPathsBefore, err := filepath.Glob(filepath.Join(mainDir, "leaf_vlog", "value-l*.log"))
+	if err != nil {
+		t.Fatalf("glob leaf_vlog before vacuum: %v", err)
+	}
+	if len(leafPathsBefore) == 0 {
+		t.Fatalf("expected leaf_vlog files before vacuum")
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/maintenance_leaf_vlog_test.go:195]: Fixed in 4552c16b: TestVacuumIndexOffline_LeafPagesInValueLog_ReopenParity now asserts each pre-vacuum leaf_vlog segment path is removed after offline vacuum.

File: cmd/collection_canonical_bench/main.go [hunk 9]
```diff
@@ -26,7 +26,7 @@ const runDirSentinel = ".collection_canonical_bench_run"
 const (
 	phasePostInsert               = "post_insert"
 	phaseOnlineOnePassMaintenance = "online_one_pass_maintenance"
-	phaseOfflineRewrite           = "offline_rewrite"
+	phaseOfflineRewrite           = "offline_compact"
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/collection_canonical_bench/main.go]: Fixed in 4552c16b: renamed phaseOfflineRewrite to phaseOfflineCompact while keeping the serialized phase value offline_compact.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@codex review

File: TreeDB/db/db.go [hunk 10]
```diff
@@ -1431,6 +1431,22 @@ func openWithLock(opts Options, lock *lockfile.Lock) (*DB, error) {
 		return nil, err
 	}
 	if opts.IndexOuterLeavesInValueLog {
+		reader := newValueReader(vm.CurrentSet())
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/db.go]: Fixed in 4552c16b and retained in 5af6689c/735c201a: the temporary CurrentSet is now released after the recovery scan, including the reset path.

File: TreeDB/db/db.go [hunk 11]
```diff
@@ -1430,6 +1430,22 @@ func openWithLock(opts Options, lock *lockfile.Lock) (*DB, error) {
...
+	recoveredLeafReset, err := recoverLeafGenerationResetAfterOfflineVacuum(opts.Dir, p, reader, db.meta.UserRootPageID, db.meta.SystemRootPageID, db.meta.CommitSeq)
+	if recoverySet != nil {
+		vm.Release(recoverySet)
+	}
+	if err != nil {
+		db.Close()
+		return nil, err
+	}
+	if recoveredLeafReset {
+		if err := vm.Refresh(); err != nil {
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/db.go:1452]: Fixed in 5af6689c: open-time recovery now releases the temporary value-log snapshot and calls vm.EvictSegment for each stale leaf_vlog file before deleting it, so the live manager no longer retains removed leaf segment handles.

File: TreeDB/db/db.go [hunk 12]
```diff
@@ -1430,6 +1430,22 @@ func openWithLock(opts Options, lock *lockfile.Lock) (*DB, error) {
...
+			db.Close()
+			return nil, err
+		}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/db.go:1455]: Fixed in 735c201a: recoverLeafGenerationResetAfterOfflineVacuum now returns true only when the reset actually runs; marker-only cleanup with live leaf refs no longer triggers the extra refresh path.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@codex review

File: TreeDB/db/vacuum_offline.go [hunk 13]
```diff
@@ -183,6 +201,13 @@ func vacuumIndexOffline(opts Options, fail vacuumFailpoint) error {
 		_ = d.Close()
 		return err
 	}
+	if acquiredValueLogSet != nil {
+		if err := d.valueLogManager.Release(acquiredValueLogSet); err != nil {
+			_ = d.Close()
+			return err
+		}
+		acquiredValueLogSet = nil
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/vacuum_offline.go]: Fixed in 83845a31: the explicit release path now clears acquiredValueLogSet before calling Release, so the deferred cleanup cannot release the same set again if Release returns an error.

File: TreeDB/db/db.go [hunk 14]
```diff
@@ -1430,6 +1430,30 @@ func openWithLock(opts Options, lock *lockfile.Lock) (*DB, error) {
 		db.Close()
 		return nil, err
 	}
+	recoverySet := vm.CurrentSet()
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/db.go]: Fixed in 83845a31: after the checked vm.Refresh call, recovery now uses CurrentSetNoRefresh so open does not perform a redundant scan with discarded errors.

**Comment by snissn** [comment]
@codex review
